scripts/state_space_models_lib.py

```diff
@@ -0,0 +1,316 @@
+        """
+        Run the backward pass in Kalman smoother.
+
+        The backward smoothing is discussed in section 10.4.4 of Kevin P. Murphy, 2022,
+        Probabilistic Machine Learning: Advanced Topics
+        The inputs are returned by forward_filter function
+
+        Parameters
+        ----------
+        * array(*batch_size, timesteps, state_size):
+            Filtered means mut
+        * array(*batch_size, timesteps, state_size, state_size)
+            Filtered covariances Sigmat
+        * array(*batch_size, timesteps, state_size)
+            Filtered conditional means mut|t-1
+        * array(*batch_size, timesteps, state_size, state_size)
+            Filtered conditional covariances Sigmat|t-1
+
+        Returns
+        -------
+        * array(*batch_size, timesteps, state_size)
+            Means of the smoothed marginal distributions
+        * array(*batch_size, timesteps, state_size, state_size)
+            Covariances of the smoothed marginal distributions
+        """
+        *batch_shape, timesteps, _ = filtered_means.shape
+        state_mean_dims = (*batch_shape, timesteps, self.state_size)
+        state_cov_dims = (*batch_shape, timesteps, self.state_size, self.state_size)
+
+        filtered_means = filtered_means.reshape(-1, timesteps, self.state_size)
+        filtered_covs = filtered_covs.reshape(-1, timesteps, self.state_size, self.state_size)
+        mu_cond_hist = mu_cond_hist.reshape(-1, timesteps, self.state_size)
+        Sigma_cond_hist = Sigma_cond_hist.reshape(-1, timesteps, self.state_size, self.state_size)
+
+        backward_map = jax.vmap(self.__backward_smoothing_pass, 0)
+        smoothed_means, smoothed_covs = backward_map(filtered_means, filtered_covs, mu_cond_hist, Sigma_cond_hist)
+
+        smoothed_means = smoothed_means.reshape(state_mean_dims)
+        smoothed_covs = smoothed_covs.reshape(state_cov_dims)
+
+        return smoothed_means, smoothed_covs
+    
+    def posterior_marginals(self, x: Array) -> Array:
+        """
+        Run a Kalman smoother to return posterior mean and cov.
+        This function only performs smoothing. If the user wants the intermediate values,
+        which are returned by filtering pass `forward_filter`, one could get it by
+
+        Parameters
+        ----------
+        x_hist: array(*batch_size, timesteps, observation_size)
+
+        Returns
+        -------
+        * array(*batch_size, timesteps, state_size)
+            Means of the smoothed marginal distributions
+        * array(*batch_size, timesteps, state_size, state_size)
+            Covariances of the smoothed marginal distributions
+        """
+        filtered_means, filtered_covs, mu_cond_hist, Sigma_cond_hist = self.forward_filter(x)
+        smoothed_means, smoothed_covs = self.backward_smoothing_pass(filtered_means, filtered_covs, mu_cond_hist, Sigma_cond_hist)
+        return smoothed_means, smoothed_covs
+
+    def log_prob(self, x: Array) -> Array:
+        """
```

Comment:
This code is very hard to read. You should compute Z(t) at each step from the normalization constant of the KF step. Then you can just sum  log Z across time. Logprob method can then just use output of the filtering / forwards algorithm.

(In the equation below I am missing an = sign)

![Screen Shot 2021-07-30 at 3 06 22 PM](https://user-images.githubusercontent.com/4632336/127716145-8ec64e28-4e15-440d-a6de-19a12b6363b3.png)